File: funnyvote/app/src/test/java/com/heaton/funnyvote/ui/account/AccountPresenterTest.java
package com.heaton.funnyvote.ui.account;

import com.heaton.funnyvote.data.user.UserDataRepository;
import com.heaton.funnyvote.database.User;
import com.heaton.funnyvote.database.VoteData;
import com.heaton.funnyvote.utils.schedulers.BaseSchedulerProvider;
import com.heaton.funnyvote.utils.schedulers.ImmediateSchedulerProvider;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyBoolean;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class AccountPresenterTest {

    private static User user;
    private static final String voteCode = "CODE_123";
    private static final VoteData voteData = new VoteData();
    private static List<VoteData> voteDataList = new ArrayList<>();

    private AccountPresenter presenter;

    @Mock
    private AccountContract.View view;

    @Mock
    private UserDataRepository userDataRepository;
    private BaseSchedulerProvider schedulerProvider;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        schedulerProvider = new ImmediateSchedulerProvider();
        user = mock(User.class);
        when(user.getUserName()).thenReturn("Heaton");
    }

    @Test
    public void createPresenter_setsThePresenterToView() {
        // Get a reference to the class under test
        presenter = new AccountPresenter(userDataRepository
                , view, schedulerProvider);

        // Then the presenter is set to the userPageView
        verify(view).setPresenter(presenter);
    }

    @Test
    public void getUserFromRepositoryAndUpdateToView() {
        when(userDataRepository.getUser(eq(false))).thenReturn(rx.Observable.just(user));
        presenter = new AccountPresenter(userDataRepository
                , view, schedulerProvider);
        presenter.subscribe();
        verify(userDataRepository).getUser(eq(false));
        verify(view).showUser(any(User.class));
    }

    @Test
    public void logoutUserAndUpdateToView() {
        when(userDataRepository.getUser(eq(false))).thenReturn(rx.Observable.just(user));
        presenter = new AccountPresenter(userDataRepository
                , view, schedulerProvider);

        when(user.getType()).thenReturn(User.TYPE_FACEBOOK);
        presenter.setUser(user);
        presenter.logout();
        verify(view).facebookLogout();

        when(user.getType()).thenReturn(User.TYPE_GOOGLE);
        presenter.setUser(user);
        presenter.logout();
        verify(view).googleSignOut();

        when(user.getType()).thenReturn(User.TYPE_TWITTER);
        presenter.setUser(user);
        presenter.logout();
        verify(view).twitterlogout();
    }

    @Test
    public void loginUserAndUpdateToView() {
        presenter = new AccountPresenter(userDataRepository
                , view, schedulerProvider);
        presenter.login(AccountPresenter.LOGIN_FB, true);
        verify(view).facebookLogin();

        presenter.login(AccountPresenter.LOGIN_GOOGLE, true);
        verify(view).googleSignIn();

        presenter.login(AccountPresenter.LOGIN_TWITTER, true);
        verify(view).twitterLogin();
    }

    @Test
    public void registerUserAndUpdateToView() {
        when(userDataRepository.registerUser(anyString(), any(User.class)
                , anyBoolean())).thenReturn(rx.Observable.just(true));
        when(userDataRepository.getUser(anyBoolean()))
                .thenReturn(rx.Observable.just(user));
        presenter = new AccountPresenter(userDataRepository
                , view, schedulerProvider);
        presenter.registerUser(user, "appId");
        verify(userDataRepository).registerUser(anyString(), any(User.class)
                , anyBoolean());
        //registerUserCallbackArgumentCaptor.getValue().onSuccess();
        verify(userDataRepository).getUser(anyBoolean());
        //getUserCallbackArgumentCaptor.getValue().onResponse(user);
        verify(view).showUser(any(User.class));
    }

    @Test
    public void registerUserFailureAndUpdateToView() {
        when(userDataRepository.registerUser(anyString(), any(User.class)
                , anyBoolean())).thenReturn(rx.Observable.error(new Exception("test error")));
        when(userDataRepository.getUser(anyBoolean()))
                .thenReturn(rx.Observable.error(new Exception("test error")));
        presenter = new AccountPresenter(userDataRepository
                , view, schedulerProvider);
        presenter.registerUser(user, "appId");
        verify(userDataRepository).registerUser(anyString(), any(User.class), anyBoolean());
        verify(userDataRepository).getUser(anyBoolean());
        verify(view).showLoginView(anyString());
    }

    @Test
    public void changeUserNameFailureAndUpdateToView() {
        when(userDataRepository.changeCurrentUserName(anyString()))
                .thenReturn(rx.Observable.just(user));
        when(userDataRepository.getUser(anyBoolean()))
                .thenReturn(rx.Observable.error(new Exception("test error")));
        presenter = new AccountPresenter(userDataRepository
                , view, schedulerProvider);
        presenter.changeCurrentUserName("new name");
        verify(userDataRepository).changeCurrentUserName(anyString());
        verify(userDataRepository).getUser(anyBoolean());
        verify(view).showLoginView(anyString());
    }

    @Test
    public void unregisterUser() {
        presenter = new AccountPresenter(userDataRepository
                , view, schedulerProvider);
        presenter.unregisterUser();
        verify(userDataRepository).unregisterUser();
    }

}

File: funnyvote/app/src/test/java/com/heaton/funnyvote/ui/account/AccountPresenterTest.kt
package com.heaton.funnyvote.ui.account

import com.heaton.funnyvote.any
import com.heaton.funnyvote.capture
import com.heaton.funnyvote.data.user.UserDataRepository
import com.heaton.funnyvote.data.user.UserDataSource
import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.database.VoteData
import com.heaton.funnyvote.ui.account.AccountContract.View

import org.junit.Before
import org.junit.Test
import org.mockito.*

import java.util.ArrayList

import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`

class AccountPresenterTest {

    private lateinit var presenter: AccountPresenter

    @Mock
    private lateinit var view: AccountContract.View

    @Mock
    private lateinit var userDataRepository: UserDataRepository
    @Captor
    private lateinit var getUserCallbackArgumentCaptor: ArgumentCaptor<UserDataSource.GetUserCallback>
    @Captor
    private lateinit var registerUserCallbackArgumentCaptor: ArgumentCaptor<UserDataSource.RegisterUserCallback>
    @Captor
    private lateinit var changeUserNameCallbackArgumentCaptor: ArgumentCaptor<UserDataSource.ChangeUserNameCallback>

    @Before
    @Throws(Exception::class)
    fun setUp() {
        MockitoAnnotations.initMocks(this)

        user = mock(User::class.java)
        `when`(user.getUserName()).thenReturn("Heaton")
    }

    @Test
    fun createPresenter_setsThePresenterToView() {
        // Get a reference to the class under test
        presenter = AccountPresenter(userDataRepository, view)

        // Then the presenter is set to the userPageView
        verify<View>(view).setPresenter(presenter)
    }

    @Test
    fun getUserFromRepositoryAndUpdateToView() {
        presenter = AccountPresenter(userDataRepository, view)
        presenter.start()
        verify(userDataRepository).getUser(capture(getUserCallbackArgumentCaptor), ArgumentMatchers.eq(false))
        getUserCallbackArgumentCaptor.value.onResponse(user)
        verify<View>(view).showUser(any<User>())
    }

    @Test
    fun logoutUserAndUpdateToView() {
        presenter = AccountPresenter(userDataRepository, view)

        `when`(user.type).thenReturn(User.TYPE_FACEBOOK)
        presenter.user = user
        presenter.logout()
        verify<View>(view).facebookLogout()

        `when`(user.type).thenReturn(User.TYPE_GOOGLE)
        presenter.user = user
        presenter.logout()
        verify<View>(view).googleSignOut()

        `when`(user.type).thenReturn(User.TYPE_TWITTER)
        presenter.user = user
        presenter.logout()
        verify<View>(view).twitterlogout()
    }

    @Test
    fun loginUserAndUpdateToView() {
        presenter = AccountPresenter(userDataRepository, view)
        presenter.login(AccountPresenter.LOGIN_FB, true)
        verify<View>(view).facebookLogin()

        presenter.login(AccountPresenter.LOGIN_GOOGLE, true)
        verify<View>(view).googleSignIn()

        presenter.login(AccountPresenter.LOGIN_TWITTER, true)
        verify<View>(view).twitterLogin()
    }

    @Test
    fun registerUserAndUpdateToView() {
        presenter = AccountPresenter(userDataRepository, view)
        presenter.registerUser(user, "appId")
        verify(userDataRepository).registerUser(ArgumentMatchers.anyString(), any<User>(), ArgumentMatchers.anyBoolean(), capture(registerUserCallbackArgumentCaptor))
        registerUserCallbackArgumentCaptor.value.onSuccess()
        verify(userDataRepository).getUser(capture(getUserCallbackArgumentCaptor), ArgumentMatchers.anyBoolean())
        getUserCallbackArgumentCaptor.value.onResponse(user)
        verify<View>(view).showUser(any<User>())
    }

    @Test
    fun registerUserFailureAndUpdateToView() {
        presenter = AccountPresenter(userDataRepository, view)
        presenter.registerUser(user, "appId")
        verify(userDataRepository).registerUser(ArgumentMatchers.anyString(), any<User>(), ArgumentMatchers.anyBoolean(), capture(registerUserCallbackArgumentCaptor))
        registerUserCallbackArgumentCaptor.value.onFailure()
        verify(userDataRepository).getUser(capture(getUserCallbackArgumentCaptor), ArgumentMatchers.anyBoolean())
        getUserCallbackArgumentCaptor.value.onFailure()
        verify<View>(view).showLoginView(ArgumentMatchers.anyString())
    }

    @Test
    fun changeUserNameFailureAndUpdateToView() {
        presenter = AccountPresenter(userDataRepository, view)
        presenter.changeCurrentUserName("new name")
        verify(userDataRepository).changeCurrentUserName(ArgumentMatchers.anyString(), capture(changeUserNameCallbackArgumentCaptor))
        changeUserNameCallbackArgumentCaptor.value.onSuccess()
        verify(userDataRepository).getUser(capture(getUserCallbackArgumentCaptor), ArgumentMatchers.anyBoolean())
        getUserCallbackArgumentCaptor.value.onFailure()
        verify<View>(view).showLoginView(ArgumentMatchers.anyString())
    }

    @Test
    fun unregisterUser() {
        presenter = AccountPresenter(userDataRepository, view)
        presenter.unregisterUser()
        verify(userDataRepository).unregisterUser()
    }

    companion object {

        private lateinit var user: User
        private val voteCode = "CODE_123"
        private val voteData = VoteData()
        private val voteDataList = ArrayList<VoteData>()
    }

}
